sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp

```diff
@@ -104,38 +157,64 @@ template <typename Ty> Ty atomic_max(Ty *ptr, Ty val) {
   // TODO: Windows will be supported soon
   __ESIMD_UNSUPPORTED_ON_HOST;
 #else
-  // TODO FIXME: fix implementation for FP types.
   if constexpr (std::is_integral_v<Ty>) {
     Ty _old, _new;
     do {
       _old = *ptr;
       _new = std::max<Ty>(_old, val);
     } while (!__atomic_compare_exchange_n(ptr, &_old, _new, false,
                                           __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST));
-    return _new;
+    return _old;
   } else {
-    __ESIMD_UNSUPPORTED_ON_HOST;
+    Ty _old, _new;
+    CmpxchgTy<Ty> _old_bits, _new_bits;
+    do {
+      _old = *ptr;
+      _new = std::max(_old, val);
+      _old_bits = *(CmpxchgTy<Ty> *)&_old;
+      _new_bits = *(CmpxchgTy<Ty> *)&_new;
+    } while (!__atomic_compare_exchange_n((CmpxchgTy<Ty> *)(CmpxchgTy<Ty> *)ptr,
+                                          &_old_bits, _new_bits, false,
+                                          __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST));
+    return _old;
   }
 #endif
 }
 
-template <typename Ty> Ty atomic_cmpxchg(Ty *ptr, Ty expected, Ty desired) {
+template <typename Ty> Ty atomic_cmpxchg(Ty *ptr, Ty desired, Ty expected) {
 #ifdef _WIN32
   // TODO: Windows will be supported soon
   __ESIMD_UNSUPPORTED_ON_HOST;
 #else
-  // TODO FIXME: fix implementation for FP types.
   if constexpr (std::is_integral_v<Ty>) {
-    Ty _old = expected;
-    __atomic_compare_exchange_n(ptr, &_old, desired, false, __ATOMIC_SEQ_CST,
+    Ty local = expected;
+    __atomic_compare_exchange_n(ptr, &local, desired, false, __ATOMIC_SEQ_CST,
                                 __ATOMIC_SEQ_CST);
-    return *ptr;
+    // if exchange occured, this means 'local=expected=*ptr'. So local
+    // is returned as old val
+    // if exchange did not occur, *ptr value compared against 'local'
+    // is stored in 'local'. So local is returned as old val
+    return local;
   } else {
-    __ESIMD_UNSUPPORTED_ON_HOST;
+    CmpxchgTy<Ty> desired_bits = *(CmpxchgTy<Ty> *)&desired;
+    CmpxchgTy<Ty> local_bits = *(CmpxchgTy<Ty> *)&expected;
+    __atomic_compare_exchange_n((CmpxchgTy<Ty> *)ptr, &local_bits, desired_bits,
+                                false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+    return *((Ty *)&local_bits);
   }
 #endif
 }
 
+// Template argument is to avoid failure from 'odr.cpp' test
```

Comment:
I think this is a bad interface - using template parameter, please fix.

If ODR fails, a fix should be different.

Reply:
Template argument is replaced with `inline`.